### PR TITLE
Enable DelayedDispatch in PEI

### PR DESCRIPTION
**Pull request by Javagedes**
## Description

Allow callbacks in PEI to register for minimum delay. PEI dispatch will continue to loop if there are any outstanding DelayDispatch registrations.

Includes:
https://github.com/microsoft/mu_basecore/commit/c5c17265162b7d97123e2f6a4be21cb48ced2eff
https://github.com/microsoft/mu_basecore/commit/30db3093fe2f7bee11422010be45f652966f1366
https://github.com/microsoft/mu_basecore/commit/dd55bfd (a portion)
https://github.com/microsoft/mu_basecore/commit/15d58b91ff21243c6d5dab99b3982aec582517ae

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

<_Please describe the test(s) that were run to verify the changes._>

## Integration Instructions

<_Describe how these changes should be integrated. Use N/A if nothing is required._>

### CODE REVIEW

**Review by Flickdm — APPROVED** [review]

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c [hunk 1]
```diff
@@ -5,9 +5,325 @@ Copyright (c) 2006 - 2024, Intel Corporation. All rights reserved.<BR>
...
+BOOLEAN
+DelayedDispatchDispatcher (
+  IN DELAYED_DISPATCH_TABLE  *DelayedDispatchTable,
+  IN EFI_GUID                *UniqueId           OPTIONAL
+  );
+
+/**
+ * Delayed Dispatch Notify function for EndoPei
+ *
+ * @return EFI_STATUS EFIAPI
```
Comment (makubacki) [inline MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: Can you please clean up the return status? `EFIAPI` doesn't make sense.

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c [hunk 2]
```diff
@@ -5,9 +5,325 @@ Copyright (c) 2006 - 2024, Intel Corporation. All rights reserved.<BR>
...
+ * GetDelayedDispatchTable
+ *
+ * @param  None
+ *
+ *
+ * @return DELAYED_DISPATCH_TABLE*
+ */
+DELAYED_DISPATCH_TABLE *
+GetDelayedDispatchTable (
+  )
```
Comment (makubacki) [inline MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]:
```suggestion
GetDelayedDispatchTable (
  VOID
  )

```

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c [hunk 3]
```diff
@@ -5,9 +5,325 @@ Copyright (c) 2006 - 2024, Intel Corporation. All rights reserved.<BR>
...
+};
+
+/**
+ * GetDelayedDispatchTable
+ *
+ * @param  None
+ *
+ *
+ * @return DELAYED_DISPATCH_TABLE*
+ */
```
Comment (makubacki) [inline MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]:
```suggestion
/**
 * GetDelayedDispatchTable
 *
 * @return DELAYED_DISPATCH_TABLE*
 */
```

Technically, these should drop the asterisks to match the latest comment style, but I don't think that needs to be changed here.

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c [hunk 4]
```diff
@@ -5,9 +5,325 @@ Copyright (c) 2006 - 2024, Intel Corporation. All rights reserved.<BR>
...
+  IN  EFI_GUID                       *UniqueId   OPTIONAL,
+  IN  UINT32                         Delay
+  )
+{
+  DELAYED_DISPATCH_TABLE  *DelayedDispatchTable;
+  DELAYED_DISPATCH_ENTRY  *Entry;
+
+  // Check input parameters
+  if ((NULL == Function) || (Delay > FixedPcdGet32 (PcdDelayedDispatchMaxDelayUs)) || (NULL == This)) {
+    DEBUG ((DEBUG_ERROR, " %a - Invalid parameter\n", __FUNCTION__));
```
Comment (makubacki) [inline MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: We've been trying to convert these to `__func__` as changes are added, but feel free to ignore.

```suggestion
    DEBUG ((DEBUG_ERROR, " %a - Invalid parameter\n", __func__));
```

File: MdeModulePkg/Include/Ppi/DelayedDispatch.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  PPI that allows a Pei module to request a callback after a minumum
+  delay.
+
+  Copyright (c) Microsoft Corporation. All rights reserved.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef __EFI_DELAYED_DISPATCH_PPI_H__
+#define __EFI_DELAYED_DISPATCH_PPI_H__
```
Comment (makubacki) [inline MdeModulePkg/Include/Ppi/DelayedDispatch.h]:
```suggestion
#ifndef EFI_DELAYED_DISPATCH_PPI_H_
#define EFI_DELAYED_DISPATCH_PPI_H_

```

File: MdeModulePkg/Include/Ppi/DelayedDispatch.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
+/**
+
+  PPI that allows a Pei module to request a callback after a minumum
+  delay.
+
+  Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (makubacki) [inline MdeModulePkg/Include/Ppi/DelayedDispatch.h]:
```suggestion
  Copyright (c) Microsoft Corporation.
```

**Review by makubacki — APPROVED** [review]

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c (hunk 1 above)
Comment (Javagedes) [reply MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: Updated to `@return EFI_STATUS`

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c (hunk 3 above)
Comment (Javagedes) [reply MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: removed @param None as requested.

File: MdeModulePkg/Include/Ppi/DelayedDispatch.h (hunk 5 above)
Comment (Javagedes) [reply MdeModulePkg/Include/Ppi/DelayedDispatch.h]: updated to `EFI_DELAYED_DISPATCH_PPI_H_` as requested.

File: MdeModulePkg/Include/Ppi/DelayedDispatch.h (hunk 6 above)
Comment (Javagedes) [reply MdeModulePkg/Include/Ppi/DelayedDispatch.h]: Removed as requested

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c (hunk 4 above)
Comment (Javagedes) [reply MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: Updated all to __func__ as requested

File: MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c (hunk 2 above)
Comment (Javagedes) [reply MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c]: Updated as requested.

**Pull request merged**